[HUDI-3172] Refactor hudi existing modules to make more code reuse in V2 Implementation

leesf opened the pull request:
## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contribute/how-to-contribute before opening a pull request.*

## What is the purpose of the pull request

* Introduce hudi-spark3-common and hudi-spark2-common modules to place classes that would be reused in different spark versions, also introduce hudi-spark3.1.x to support spark 3.1.x
* Introduce `hudi` format under hudi-spark2, hudi-spark3, hudi-spark3.1.x modules and change the `hudi` format in original hudi-spark module to `hudi_v1` format 
* Manually tested on Spark 3.1.2 and Spark 3.2.0 SQL.

## Brief change log

*(for example:)*
  - *Modify AnnotationLocation checkstyle rule in checkstyle.xml*

## Verify this pull request

*(Please pick either of the following options)*

This pull request is a trivial rework / code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

*(example:)*

  - *Added integration tests for end-to-end.*
  - *Added HoodieClientWriteTest to verify the change.*
  - *Manually verified the change by running a job locally.*

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

leesf commented:
@hudi-bot run azure

vinothchandar commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/HoodieSqlCommonUtils.scala:
```diff
@@ -0,0 +1,316 @@
+    }
+  }
+
+  def getTableIdentify(table: LogicalPlan): TableIdentifier = {
```
getTableIdentifier?

leesf replied:
done

leesf replied:
done

vinothchandar commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/HoodieSqlCommonUtils.scala:
```diff
@@ -0,0 +1,316 @@
+
+import scala.collection.immutable.Map
+
+object HoodieSqlCommonUtils extends SparkAdapterSupport {
```
guess all of the code is moved from some place else? or these are new methods? Could you clarify? 

Can we call this class HoodieSqlAdapterSupport or something or break the `AdapterSupport` methods into a separate objecT?

leesf replied:
yes, the code is moved from HoodieSqlUtils to make more reuse and no new methods added. Also I intend to call it as a *Utils class  to keep align with HoodieSqlUtils which also extends SparkAdapterSupport.

vinothchandar commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/hudi/command/ValidateDuplicateKeyPayload.scala:
```diff
@@ -0,0 +1,45 @@
+ * Validate the duplicate key for insert statement without enable the INSERT_DROP_DUPS_OPT
+ * config.
+ */
+class ValidateDuplicateKeyPayload(record: GenericRecord, orderingVal: Comparable[_])
```
assuming this class is just moved over?

leesf replied:
yes, moved from InsertIntoHoodieTableCommand

vinothchandar reviewed (commented):
Let me make another pass at all the pom changes. That seems to be main thing here. 
In the meantime, could you clarify these comments?

Also have you tested these changes across spark 2.x and 3.1/3.2 bundles ?

xushiyan commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/DefaultSource.scala:
```diff
@@ -177,7 +177,7 @@ class DefaultSource extends RelationProvider
   }
 
-  override def shortName(): String = "hudi"
+  override def shortName(): String = "hudi_v1"
```
@leesf i suppose this refactoring PR not meant to include this change?

leesf replied:
If not change the format, it would conflict with hudi-spark2/hudi-spark3.1.x/hudi-spark3 module format.

vinothchandar replied:
would it conflict? Given we are extending `DefaultSource` and overriding `shortName()`?

leesf replied:
it is because in hudi-spark-bundle module. I used `<transformer implementation="org.apache.maven.plugins.shade.resource.AppendingTransformer">
                  <resource>META-INF/services/org.apache.spark.sql.sources.DataSourceRegister</resource>
                </transformer>` to append the formats(hudi_v1 and hudi) in `DataSourceRegister` file, so it will conflict if not change the format. As to the PR itself, we do not need to change the format to `hudi_v1` and not use AppendingTransformer. But when implementing V2 codepath, I find it difficult to handle the incremental bootstrap table(https://github.com/apache/hudi/blob/ac8d014a0602e3c499771f3313f0f88de57cdda1/hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/IncrementalRelation.scala#L174) as it will pass schema to Spark and difficult to handle in v2 codepath, after finding a good way to handle it, we would definitely delete the hudi_v1 format here.

leesf commented:
> Let me make another pass at all the pom changes. That seems to be main thing here. In the meantime, could you clarify these comments?
> 
> Also have you tested these changes across spark 2.x and 3.1/3.2 bundles ?

@vinothchandar Yes, I have manually tested it with spark 3.2.0 and spark 3.1.2 version on spark sql. and the CI tested it on spark 2.4.x and works well.

xushiyan reviewed (commented):
LGTM. Since a lot of files were moved, cherry pick later commits for 0.10.1 could be problematic. Shall we move this into a feature branch?

vinothchandar commented:
@xushiyan given we are almost winding down for 0.10.1, I suggest we land this sooner than later. That way we can focus on stabilizing master for 0.11.0 - that's not too far away. wdyt?

xushiyan commented:
> @xushiyan given we are almost winding down for 0.10.1, I suggest we land this sooner than later. That way we can focus on stabilizing master for 0.11.0 - that's not too far away. wdyt?

it won't be too far away. @nsivabalan is cherry picking for 0.10.1 which will complete by Jan 9. Holding this off for 2 more days can avoid conflicts from some Spark fixes merged after this. Seeing 2 more fixes coming. After that we should be able to land this right away.

vinothchandar commented:
2 more days should be okay?

vinothchandar commented:
@leesf few questions. 

1. What do the hudi-spark2-extensions/hudi-spark3-extensions do? What code would these have in the future?

2. Users may have spark jobs that depend directly on `hudi-spark` . any impact to these from the changes?

3. I feel there are way too many packages now - hudi-spark*, then it depends hudi-spark-common* . Is there a way for us to merge hudi-spark-common into hudi-spark* ?  is the hudi-spark3-common so we can share code across the different 3.x versions? Can you add a README that explains what code should go where going forward

vinothchandar reviewed (approved):
Made some cleanup suggestions. LGTM overall

nsivabalan commented:
yeah, would really appreciate if we can wait until Jan 9 to land this patch. thanks!

leesf commented:
> @leesf few questions.
> 
> 1. What do the hudi-spark2-extensions/hudi-spark3-extensions do? What code would these have in the future?
> 2. Users may have spark jobs that depend directly on `hudi-spark` . any impact to these from the changes?
> 3. I feel there are way too many packages now - hudi-spark*, then it depends hudi-spark-common* . Is there a way for us to merge hudi-spark-common into hudi-spark* ?  is the hudi-spark3-common so we can share code across the different 3.x versions? Can you add a README that explains what code should go where going forward
> 4. Any updates to release notes or README?

1. I did intend to remove the modules but wrongly added. removed now.
2. I did not change hudi-spark module, so it should have no impact on the users.
3. I think the current modules has be in a good shape, if merge code from hudi-spark-common into hudi-spark* or merge code from hudi-spark3-common into hudi-spark3 & spark3.1.x, there would be many duplicate classes, which i think is not an ideal way.
4. Added a README under hudi-spark-datasource module.

leesf commented:
> yeah, would really appreciate if we can wait until Jan 9 to land this patch. thanks!

@nsivabalan @xushiyan time to land this patch?

xushiyan commented:
@leesf yes we won't have conflicting patches to pick from master. we can land this one now.

leesf commented:
> @leesf yes we won't have conflicting patches to pick from master. we can land this one now.

@xushiyan CI passed, so can we merge?

leesf commented:
@hudi-bot run azure

leesf commented:
@hudi-bot run azure

leesf commented:
@hudi-bot run azure

leesf commented:
@hudi-bot run azure

leesf commented:
@hudi-bot run azure

leesf commented:
@hudi-bot run azure
